### ISSUE TITLE
Implement priority queue as an option for reconcile loops

### ISSUE THREAD

**Issue by fabriziopandini** [kubernetes-sigs/controller-runtime#2374]
While discussing https://github.com/kubernetes-sigs/controller-runtime/issues/857#issuecomment-1587149384 an idea come up about having something like a priority queue, so we can give resync events a lower priority while assigning to other events a higher priority.

This could help when there are many objects of the same type and at every resync period there is a storm of events being added to the queue.

cc @alvaroaleman @sbueringer

**Comment by alvaroaleman** [comment]
While resync events are one example, I think that it makes sense in general to allow specifying a priority when adding items to the workqueue.

One challenge here is the fact that as of today, there is no priority queue implementation in upstream kube.

**Comment by alvaroaleman** [comment]
All,

I've written a small [design doc](https://github.com/kubernetes-sigs/controller-runtime/pull/3013) and [POC](https://github.com/kubernetes-sigs/controller-runtime/pull/3014) for this. Would appreciate everyones feedback

**Comment by sbueringer** [comment]
I've updated the issue so we can use it as an umbrella issue for the feature going forward

**Comment by lingdie** [comment]
Can we provide more example codes on priority queue feature?

**Comment by lingdie** [comment]
Is that all? Can't believe. :)

**Comment by sbueringer** [comment]
Yup. Literally just that. We try to implement this in a way that it just works

**Comment by vishalbalaji-v** [comment]
Hi everyone,

Is there timeline on when this will be generally available? The doc for priority queue says it's currently in beta ([link](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/controller/controller.go#L91-L95)). 

We have a use case for it, but people are worried about putting it in production since the doc says it's in beta. I'd be happy to contribute to something if required. The [design](https://github.com/kubernetes-sigs/controller-runtime/blob/main/designs/priorityqueue.md) also doesn't seem to have been updated in a while.

**Comment by alvaroaleman** [comment]
> Is there timeline on when this will be generally available? The doc for priority queue says it's currently in beta ([link](https://github.com/kubernetes-sigs/controller-runtime/blob/main/pkg/controller/controller.go#L91-L95)).

The main reason for that is that we don't want to call it GA until we are sure we don't want to change the interface definition. The priority queue is being used in production in places where it matters and there are no known issues with it.

**Comment by sbueringer** [comment]
And PQ is enabled per default 🎉 

Thx to everyone that is contributing to this feature!

**Comment by fabriziopandini** [comment]
Awesome job folks!
